cmd/generateproxy/interceptor.go

```diff
@@ -0,0 +1,377 @@
+		if resolved == f.String() {
+			return true
+		}
+	}
+
+	return false
+}
+
+// resolveType returns the underlying type for pointers, slices and maps
+func resolveType(t types.Type) types.Type {
+	switch typ := t.(type) {
+	case *types.Pointer:
+		return resolveType(typ.Elem())
+	case *types.Slice:
+		return resolveType(typ.Elem())
+	case *types.Map:
+		return resolveType(typ.Elem())
+	}
+
+	return t
+}
+
+func pruneRecords(input map[string]*TypeRecord) map[string]*TypeRecord {
+	result := map[string]*TypeRecord{}
+
+	for typ, record := range input {
+		if record.Matches {
+			result[typ] = record
+		}
+	}
+
+	return result
+}
+
+// walk iterates the methods on a type and returns whether any of them can eventually lead to Payload(s)
+// The return type for each method on this type is walked recursively to decide which methods can lead to Payload(s)
+func walk(desired []types.Type, typ types.Type) bool {
+	// If this type is a slice then walk the underlying type and then make a note we need to encode slices of this type
+	if isSlice(typ) {
+		result := walk(desired, elemType(typ))
+		record := records[typeName(typ)]
+		record.Slice = true
+		return result
+	}
+
+	// If this type is a map then walk the underlying type and then make a note we need to encode maps with values of this type
+	if isMap(typ) {
+		result := walk(desired, elemType(typ))
+		record := records[typeName(typ)]
+		record.Map = true
+		return result
+	}
+
+	typeName := typeName(typ)
+
+	// If we've walked this type before, return the previous result
+	if record, ok := records[typeName]; ok {
+		return record.Matches
+	}
+
+	record := TypeRecord{}
+	records[typeName] = &record
+
+	for _, meth := range typeutil.IntuitiveMethodSet(elemType(typ), nil) {
+		// Ignore non-exported methods
+		if !meth.Obj().Exported() {
+			continue
+		}
+
+		methodName := meth.Obj().Name()
+		// The protobuf types have a Get.. method for every protobuf they refer to
+		// We walk only these methods to avoid cycles or other nasty issues
+		if !strings.HasPrefix(methodName, "Get") {
+			continue
+		}
+
+		sig := meth.Obj().Type().(*types.Signature)
+		// All the Get... methods return the relevant protobuf as the first result
+		resultType := sig.Results().At(0).Type()
+
+		// Check if this method returns a Payload(s) or if it leads (eventually) to a Type which refers to a Payload(s)
+		if typeMatches(resultType, desired...) || walk(desired, resultType) {
+			record.Matches = true
+			record.Methods = append(record.Methods, methodName)
+		}
+	}
+
+	// Return whether this Type can (eventually) lead to Payload(s)
+	// This is used in the walk logic above so that our encoding handles intermediate Types between our Request/Response objects and Payload(s)
+	return record.Matches
+}
+
+func generateInterceptor(cfg config) {
+	conf := &packages.Config{Mode: packages.NeedImports | packages.NeedTypes | packages.NeedTypesInfo}
+	pkgs, err := packages.Load(conf, "go.temporal.io/api/workflowservice/v1")
```

Comment:
As discussed in Slack, need to have a `replace` in go.mod for this to work with latest local changes probably